src/gsym.rs

```diff
@@ -74,6 +77,93 @@ impl SymResolver for GsymResolver {
 
+    /// Finds the source code location for a given address.
+    ///
+    /// This function takes in an address and returns the file path,
+    /// line number and column of the line in the source code where
+    /// the address corresponds to. If it doesn't find any match it
+    /// returns None.
+    ///
+    /// # Arguments
+    ///
+    /// * `addr` - A 64-bit unsigned integer representing the address to find the source code location for.
+    ///
+    /// # Returns
+    ///
+    /// `Some(AddressLineInfo)` if the location is found, otherwise `None`.
+    ///
+    fn find_line_info(&self, addr: u64) -> Option<AddressLineInfo> {
+        let addr = addr - self.loaded_address;
+        let idx = find_address(&self.ctx, addr);
+        let symaddr = self.ctx.addr_at(idx);
+        if addr < symaddr {
+            return None;
+        }
+        let addrinfo = self.ctx.addr_info(idx);
+        if addr >= (symaddr + addrinfo.size as u64) {
+            return None;
+        }
+
+        let addrdatas = parse_address_data(addrinfo.data);
+        for adr_ent in addrdatas {
+            if adr_ent.typ != InfoTypeLineTableInfo {
+                continue;
+            }
+            // Continue to execute all GSYM line table operations
+            // until the end of the buffer is reached or a row
+            // containing addr is located.
+            let (lntab_hdr, hdr_bytes) = parse_line_table_header(adr_ent.data).ok()?;
+            let ops = &adr_ent.data[hdr_bytes..];
+            let mut lntab_row = line_table_row_from(&lntab_hdr, symaddr);
+            let mut last_lntab_row = lntab_row.clone();
+            let mut row_cnt = 0;
+            let mut pc = 0;
+            while pc < ops.len() {
+                match run_op(&mut lntab_row, &lntab_hdr, ops, pc) {
+                    RunResult::Ok(bytes) => {
+                        pc += bytes;
+                    }
+                    RunResult::NewRow(bytes) => {
+                        pc += bytes;
+                        row_cnt += 1;
+                        if addr < lntab_row.address {
+                            if row_cnt == 1 {
+                                // The address is lower than the first row.
+                                return None;
+                            }
+                            // Rollback to the last row.
+                            lntab_row = last_lntab_row;
+                            break;
+                        }
+                        last_lntab_row = lntab_row.clone();
+                    }
+                    RunResult::End | RunResult::Err => {
+                        break;
+                    }
+                }
+            }
+
+            if row_cnt == 0 {
+                continue;
+            }
+
+            let finfo = self.ctx.file_info(lntab_row.file_idx as usize);
+            let dirname = self.ctx.get_str(finfo.directory as usize);
+            let filename = self.ctx.get_str(finfo.filename as usize);
```

Comment:
Who sanitized these offsets?

Reply:
> `get_str()` checks the boundary just like I mentioned in the earlier comment of `file_info()`.

[No](https://github.com/libbpf/blazesym/blob/e7e4327036631dea10a992b559ca39c789e842dc/src/gsym/parser.rs#L180)? That, by the way, is exactly why we should not introduce such APIs in the first place.